mlflow/tracking/_tracking_service/client.py

```diff
@@ -886,6 +905,7 @@ def set_terminated(self, run_id, status=None, end_time=None):
         # data in the background. This call is making sure every async logging data has been
         # submitted for logging, but not necessarily finished logging.
         self.store.end_async_logging()
+        self._log_url(run_id)
```

Comment:
I'd isolate this from `set_terminated`.

Reply:
so the purpose is to ensure URL logging is called at finish time, and the `set_terminated` in `_tracking_service/client.py` is the only place I am aware that is guaranteed to be called at finishing time. Haru do you have a suggestion on alternative?

Reply:
can we just call `_log_url` after `set_terminated` is called?

Reply:
do you mean calling it in fluent API `mlflow.end_run`? Many users are not using `end_run()`.

Reply:
For example, we can log the run URL in `ActiveRun.__exit__`:

```diff
diff --git a/mlflow/tracking/fluent.py b/mlflow/tracking/fluent.py
index 7aa84ba47..b8045b975 100644
--- a/mlflow/tracking/fluent.py
+++ b/mlflow/tracking/fluent.py
@@ -195,6 +195,7 @@ class ActiveRun(Run):
 
     def __exit__(self, exc_type, exc_val, exc_tb):
         status = RunStatus.FINISHED if exc_type is None else RunStatus.FAILED
+        # log the run URL here
         end_run(RunStatus.to_string(status))
         return exc_type is None
```

Reply:
Do users always terminate runs?

```
mlflow.log_parmas(...)
```

Should this print out the run URL?

Reply:
Yea actually I thought about putting logging inside fluent API. but many users, and specifically Composer doesn't use fluent API, but use our client APIs directly to control the MLflow lifecycle, so we have to cover those people.